mods/sfinv/api.lua

```diff
@@ -0,0 +1,152 @@
+sfinv = {
+	pages = {},
+	pages_unordered = {},
+	homepage_name = "sfinv:crafting",
+	contexts = {},
+	enabled = true
+}
+
+function sfinv.register_page(name, def)
+	if not name or not def or not def.get then
+		error("Invalid sfinv page. Requires a name & def, and a get function in def")
+	end
+	sfinv.pages[name] = def
+	def.name = name
+	table.insert(sfinv.pages_unordered, def)
+end
+
+function sfinv.get_nav_fs(player, context, nav, current_idx)
+	-- Only show tabs if there is more than one page
+	if #nav > 1 then
+		return "tabheader[0,0;tabs;" .. table.concat(nav, ",") .. ";" .. current_idx .. ";true;false]"
+	else
+		return ""
+	end
+end
+
+local theme_main = "bgcolor[#080808BB;true]" .. default.gui_bg ..
+		default.gui_bg_img
+
+local theme_inv = default.gui_slots .. [[
+		list[current_player;main;0,4.7;8,1;]
+		list[current_player;main;0,5.85;8,3;8]
+	]]
+
+function sfinv.make_formspec(player, context, content, show_inv, size)
+	if not size then
+		size = "size[8,8.6]"
+	end
+	local tmp = {
+		size,
+		theme_main,
+		sfinv.get_nav_fs(player, context, context.nav_titles, context.nav_idx),
+		content
+	}
+	if show_inv then
+		tmp[#tmp + 1] = theme_inv
+	end
+	return table.concat(tmp, "")
+end
+
+function sfinv.get_formspec(player, context)
+	-- Generate navigation tabs
+	local nav = {}
+	local nav_ids = {}
+	local current_idx = 1
+	for i, pdef in pairs(sfinv.pages_unordered) do
+		if not pdef.is_in_nav or pdef:is_in_nav(player, context) then
+			nav[#nav + 1] = pdef.title
+			nav_ids[#nav_ids + 1] = pdef.name
+			if pdef.name == context.page then
+				current_idx = i
+			end
+		end
+	end
+	context.nav = nav_ids
+	context.nav_titles = nav
+	context.nav_idx = current_idx
+
+	-- Generate formspec
+	local page = sfinv.pages[context.page] or sfinv.pages["404"]
+	if page then
+		return page:get(player, context)
+	else
+		local old_page = context.page
+		context.page = sfinv.homepage_name
+		if sfinv.pages[context.page] then
+			minetest.log("warning", "[sfinv] Couldn't find " .. dump(old_page) .. " so using switching to homepage")
+			return sfinv.get_formspec(player, context)
+		else
+			error("[sfinv] Invalid homepage")
+		end
+	end
+end
+
+function sfinv.set(player, context)
+	if not context then
+		local name = player:get_player_name()
+		context = sfinv.contexts[name]
+		if not context then
+			context = {
+				page = sfinv.homepage_name
+			}
+			sfinv.contexts[name] = context
+		end
+	end
+
+	local fs = sfinv.get_formspec(player, context)
+	player:set_inventory_formspec(fs)
+end
+
+minetest.register_on_joinplayer(function(player)
```

Comment:
Wouldn't it make sense to free the contexts on leave player, to free unused memory?